signer/src/stacks/api.rs

```diff
@@ -435,6 +524,46 @@ impl StacksInteract for StacksClient {
     async fn get_tenure_info(&self) -> Result<RPCGetTenureInfo, Error> {
         self.get_tenure_info().await
     }
+    /// Estimate the high priority transaction fee for the input
+    /// transaction call given the current state of the mempool.
+    ///
+    /// This function attempts to use the POST /v2/fees/transaction
+    /// endpoint on a stacks node to estimate the current high priority
+    /// transaction fee for a given transaction. If the node does not
+    /// have enough information to provide an estimate, we then get the
+    /// current high priority fee for an STX transfer and use that as an
+    /// estimate for the transaction fee.
+    async fn estimate_fees<T>(&self, payload: &T) -> Result<u64, Error>
+    where
+        T: AsTxPayload + Send + Sync,
+    {
+        // If we cannot get an estimate for the transaction, try the
+        // generic STX transfer since we should always be able to get the
+        // STX transfer fee estimate. If that fails then we bail, maybe we
+        // should try another node.
+        let fee_estimates = self
+            .get_fee_estimate(payload)
+            .or_else(|err| async move {
+                tracing::warn!("could not estimate contract call fees: {err}");
+                // Estimating STX transfers is simple since the estimate
+                // doesn't depend on the recipient, amount, or memo. So a
+                // dummy transfer payload will do.
+                self.get_fee_estimate(&DUMMY_STX_TRANSFER_PAYLOAD).await
+            })
+            .await?;
+        // As of this writing the RPC response includes exactly 3 estimates
+        // (the low, medium, and high priority estimates). We want to use
+        // the high priority estimate (the max) to make sure that the
+        // transaction gets confirmed as quickly as possible.
+        let fee_estimate = fee_estimates
+            .estimations
+            .iter()
+            .map(|estimate| estimate.fee)
+            .max()
```

Comment:
Are we sure we want to do max and not median? One wonky API could make the fees huge.

Reply:
> Are we sure we want to do max and not median? One wonky API could make the fees huge.
> 
> If we want to do max lets make there a maximum fee config variable to safeguard against a bad API response.

I'm not certain if we should opt for the high priority fee either. On the Bitcoin side, we select the high priority fee, so I applied the same logic here. However, this analogy might not be right since the signers are the ones paying. Medium priority is more cost-effective for them, while high priority could enhance the user experience. I'm open to either approach.

Reply:
> If we want to do max lets make there a maximum fee config variable to safeguard against a bad API response.

Yeah that sounds reasonable to me. But I think that also applies to taking the medium. And we should also do that for fees on the bitcoin side. I'll open a ticket for this, we can always choose to de-prioritize it.

Reply:
Oh, we don't quite have have this issue on the bitcoin side, since we take an average of two sources.

Reply:
I opened https://github.com/stacks-network/sbtc/issues/366 for this.